Project: Change PR template and CONTRIBUTING guide

DylanGuedes opened the pull request:
**What this PR does / why we need it**:
* Move contributing instructions from the PR template to the CONTRIBUTING guide
* Rewrite the contributing guide
* Make the PR template slim by getting rid of sections that are optional but often left blank by most contributors/maintainers
* We're making the PR template slim as we're using the merged PR body as the merge commit message. Because the comments are often left there by contributors, they're prone to be merged by maintainers, hence making our commit log history dirty with a lot of undesired comments.

dannykopping reviewed (commented):
Thanks for this @DylanGuedes 
I'm concerned though that folks won't review the contributing doc before creating PRs. The template plays a vital role in reminding folks what's required, and the checklist is useful to this end.

kavirajk commented on CONTRIBUTING.md:
```diff
@@ -4,20 +4,30 @@ Loki uses GitHub to manage reviews of pull requests:
+3. Your PR is correctly documenting appropriate changes under the CHANGELOG. You should document your changes there if:
+  * It adds an important feature
+  * It fixes an issue present in a previous release
+  * It causes a change in operation that would be useful for an operator of Loki to know then please add a CHANGELOG entry.
```
```suggestion
  * It causes a change in operation that would be useful for an operator of Loki to know
```
Removing repeated mentioning of adding it in CHANGELOG as it already mentioned above it.

kavirajk commented on CONTRIBUTING.md:
```diff
@@ -4,20 +4,30 @@ Loki uses GitHub to manage reviews of pull requests:
+  * metric names or label names
+  * changes existing log lines such as the metrics.go query output line
+  * configuration parameters
+  * anything to do with any API
```
```suggestion
  * Any breaking changes to HTTP or gRPC API endpoints
```

kavirajk commented on CONTRIBUTING.md:
```diff
@@ -4,20 +4,30 @@ Loki uses GitHub to manage reviews of pull requests:
+  * It causes a change in operation that would be useful for an operator of Loki to know then please add a CHANGELOG entry.
+For documentation changes, build changes, simple fixes etc please skip this step. We are attempting to curate a changelog of the most relevant and important changes to be easier to ingest by end users of Loki.
+4. Your PR documents upgrading steps under `docs/sources/upgrading/_index.md` if it changes:
+  * default configuration values
```
Can we stick with `Sentence case` for all the bullet points here?

DylanGuedes replied:
Sure! Pushed a commit changing it.

kavirajk commented on CONTRIBUTING.md:
```diff
@@ -4,20 +4,30 @@ Loki uses GitHub to manage reviews of pull requests:
+  * It adds an important feature
+  * It fixes an issue present in a previous release
+  * It causes a change in operation that would be useful for an operator of Loki to know then please add a CHANGELOG entry.
+For documentation changes, build changes, simple fixes etc please skip this step. We are attempting to curate a changelog of the most relevant and important changes to be easier to ingest by end users of Loki.
```
```suggestion
You can skip this step for documentation changes, build related changes and simple bug fixes or enhancements. Rationale being we are attempting to curate the CHANGELOG entries with most relevant and important changes that end users of Loki care about.
```

kavirajk commented on CONTRIBUTING.md:
```diff
@@ -4,20 +4,30 @@ Loki uses GitHub to manage reviews of pull requests:
+4. Your PR documents upgrading steps under `docs/sources/upgrading/_index.md` if it changes:
+  * default configuration values
+  * metric names or label names
+  * changes existing log lines such as the metrics.go query output line
```
People outside loki squad usually won't be that familiar with `metrics.go` convention :) 
```suggestion
  * changes existing log lines that may be used in dashboard or alerts. e.g: logs lines in any `metrics.go` files might be getting used in building dashboards or alerts.
```

kavirajk reviewed (commented):
Thanks @DylanGuedes. Looks nice. left new minor suggestions.

DylanGuedes commented:
> Thanks for this @DylanGuedes I'm concerned though that folks won't review the contributing doc before creating PRs. The template plays a vital role in reminding folks what's required, and the checklist is useful to this end.

Got it. Just double checking, by reverting the checklist change would you say we'd cover this issue with folks overlooking the contributing guide?

kavirajk commented:
> I'm concerned though that folks won't review the contributing doc before creating PRs. The template plays a vital role in reminding folks what's required, and the checklist is useful to this end.

One option could be to add link to the contributing guide by some bot as a comment for every new PR (or we can add it as a link in the PR template for now). 

I personally prefer bot approach, less work during squash and merge. (but can be bit of a work I belive)

DylanGuedes commented:
> I think we can start simple and restore the checklist, and include a link to the contributing guide next to the checklist
> 
> That should give us the best of both worlds IMHO

Agreed. I returned the checklist, added an item regarding reading the contributing guide and I added a new item asking the contributor to remove the checklist after following all steps. WDYT? If you think the last one is too debatable I can remove it of this PR and let it for its own PR to concentrate discussions there and unblock this

dannykopping commented on .github/pull_request_template.md:
```diff
@@ -1,41 +1,9 @@
-3. Rebase your PR if it gets out of sync with main
--->
-
 **What this PR does / why we need it**:
```
I think we can add these sections back in, as they'd be relevant in the comments as well potentially - and they give a nice structure to the PR description.

The rest of the cruft can go to the CONTRIBUTING.md which you've done

```suggestion
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:
```

dannykopping commented on .github/pull_request_template.md:
```diff
@@ -1,41 +1,9 @@
-- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
+- [ ] `CHANGELOG.md` updated according to the CONTRIBUTING guide
 - [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`
+- [ ] Erased this checklist after completing it
```
In practice this will not likely be done.

dannykopping commented on .github/pull_request_template.md:
```diff
@@ -1,41 +1,9 @@
-Please document clearly what changed AND what needs to be done in the upgrade guide.
--->
 **Checklist**
+- [ ] Readed the `CONTRIBUTING.md` guide
```
Please link the guide

```suggestion
- [ ] Reviewed the `CONTRIBUTING.md` guide
```

dannykopping commented on .github/pull_request_template.md:
```diff
@@ -1,41 +1,9 @@
 - [ ] Documentation added
 - [ ] Tests updated
-- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
+- [ ] `CHANGELOG.md` updated according to the CONTRIBUTING guide
```
```suggestion
- [ ] `CHANGELOG.md` updated
```

We've already encouraged them to read the guide above

dannykopping reviewed (commented):
A few minor nits, sorry for the back and forth on this

DylanGuedes commented:
> A few minor nits, sorry for the back and forth on this

No worries! I've updated this with your suggestions. Only one I decided to not apply was the one regarding changing adding a link to CONTRIBUTING.md; I don't think that's worth it since the link won't be accessible from the Github editing page anyway

dannykopping reviewed (approved):
LGTM, thanks!
